design-proposals/storage-class-migration/storage-class-migration.md

```diff
@@ -0,0 +1,269 @@
+# Overview
+This design proposal extends KubeVirt APIs to enable the migration between storage classes.
+
+## Motivation
+In some circumstances, the guest owner may want to have the option of switching the storage provider or class for a group of volumes used by an active VM. The VM must currently be turned off for this operation due to service disruptions.
```

Comment:
right but either I'm missing something or users would still face service disruptions because with this design I assume the destination pod won't start until the disks are cloned to the destination storage device, right?
when the VM is running, it may write to the disk(s), so we need to pause it at the beginning?
if the goal is to provide live-storage migration, then I think we'd need to mirror the data that the VM writes to the destination storage, copy the backing volume(s), and then stop the mirroring - that way we really avoid the disruptions, otherwise the process is like hibernating the VM and resuming the VM from hibernation with disks that are placed in a different storage device, no? (yeah, instead of dumping the memory to the storage and restoring it from there, we copy it over the network which is typically faster, but users would still notice an impact on the workloads..)

Reply:
@ahadas I hope I understand your feedback correctly.

The storage migration won't cause any service disruptions. The storage is copied during the live-migration of the VM. This uses libvirt and QEMU internals and it is equivalent to the [migration with non-shared disks](https://libvirt.org/migration.html#migration-of-vms-using-non-shared-images-for-disks)

Reply:
> This uses libvirt and QEMU internals and it is equivalent to the [migration with non-shared disks](https://libvirt.org/migration.html#migration-of-vms-using-non-shared-images-for-disks)

Ah I see
@aglitke I think you're one the very few that were involved in live-storage-migration in oVirt and is still around :) why didn't we use this mechanism in oVirt, is it because we didn't want the VM to migrate to another host in the process?

Reply:
@ahadas I don't know the reason about oVirt. However, there is a more efficient mechanism that only copies the storage without live migrating the VM. I put more details in the section `Motivation on the approaches
`. There are some k8s restrictions and that's the main reason why we need to live migrate the VM.
It is possible that oVirt was using the most efficient approach rather than migrating the entire VM.

Reply:
@alicefr in oVirt there's a much more complicated process that is orchestrated by the central management unit:
1. create a snapshot (qcow) volume for the VM (on the source storage of course)
2. create a qcow volume on the destination storage
3. mirror the data that is being written by the VM to the volume the was created in step 2
4. copy the data from the backing chain of the VM on the source to a chain of volume on the destination storage
5. stop the mirroring
6. rebase the top-volume on the destination storage

now that I write this, I think the reason for that could be to preserve the (external) snapshots that exist in the source storage, but that shouldn't be an issue for KubeVirt, right?

Reply:
It isn't an issue for KubeVirt because we don't use qcow images. The disks on the PVCs are only in raw format.